src/hyperlight_host/src/hypervisor/gdb/hyp_debug.rs

```diff
@@ -0,0 +1,481 @@
+    /// The address is checked to be a valid guest address
+    fn read_addrs(
+        &mut self,
+        vcpu_fd: &Self::Vcpu,
+        mut gva: u64,
+        mut data: &mut [u8],
+        dbg_mem_access_fn: Arc<Mutex<dyn DbgMemAccessHandlerCaller>>,
+    ) -> Result<()> {
+        let data_len = data.len();
+        log::debug!("Read addr: {:X} len: {:X}", gva, data_len);
+
+        while !data.is_empty() {
+            let gpa = self.translate_gva(vcpu_fd, gva)?;
+
+            let read_len = std::cmp::min(
+                data.len(),
+                (PAGE_SIZE - (gpa & (PAGE_SIZE - 1))).try_into().unwrap(),
+            );
+            let offset = gpa as usize - SandboxMemoryLayout::BASE_ADDRESS;
+
+            dbg_mem_access_fn
+                .try_lock()
+                .map_err(|e| new_error!("Error locking at {}:{}: {}", file!(), line!(), e))?
+                .read(offset, &mut data[..read_len])?;
+
+            data = &mut data[read_len..];
+            gva += read_len as u64;
+        }
+
+        Ok(())
+    }
+
+    /// Restores the overwritten data to the guest memory
+    fn remove_sw_breakpoint(
+        &mut self,
+        vcpu_fd: &Self::Vcpu,
+        addr: u64,
+        dbg_mem_access_fn: Arc<Mutex<dyn DbgMemAccessHandlerCaller>>,
+    ) -> Result<bool> {
+        let addr = self.translate_gva(vcpu_fd, addr)?;
+
+        if self.is_sw_breakpoint(&addr) {
+            let save_data = self
+                .delete_sw_breakpoint_data(&addr)
+                .ok_or_else(|| new_error!("Expected to contain the sw breakpoint address"))?;
+
+            // Restore saved data to the guest's memory
+            self.write_addrs(vcpu_fd, addr, &save_data, dbg_mem_access_fn)?;
+
+            Ok(true)
+        } else {
+            Ok(false)
+        }
+    }
+
+    /// Copies the data from the provided slice to the guest memory address
+    /// The address is checked to be a valid guest address
+    fn write_addrs(
+        &mut self,
+        vcpu_fd: &Self::Vcpu,
+        mut gva: u64,
+        mut data: &[u8],
+        dbg_mem_access_fn: Arc<Mutex<dyn DbgMemAccessHandlerCaller>>,
+    ) -> Result<()> {
+        let data_len = data.len();
+        log::debug!("Write addr: {:X} len: {:X}", gva, data_len);
+
+        while !data.is_empty() {
+            let gpa = self.translate_gva(vcpu_fd, gva)?;
+
+            let write_len = std::cmp::min(
+                data.len(),
+                (PAGE_SIZE - (gpa & (PAGE_SIZE - 1))).try_into().unwrap(),
+            );
+            let offset = gpa as usize - SandboxMemoryLayout::BASE_ADDRESS;
+
+            dbg_mem_access_fn
+                .try_lock()
+                .map_err(|e| new_error!("Error locking at {}:{}: {}", file!(), line!(), e))?
+                .write(offset, data)?;
+
+            data = &data[write_len..];
+            gva += write_len as u64;
+        }
+
+        Ok(())
+    }
+}
+
+#[cfg(kvm)]
```

Comment:
Maybe this should be in a separate file?

Reply:
Moved to kvm specific different file